### DISCUSSION TITLE
OAL Filter expressions link

### DISCUSSION THREAD

**Discussion by BritneyStars** [apache/skywalking#7624] [Q&A] [answered]
**_The expressions support linking by and, or and (...)._** The OPs support ==, !=, >, <, >=, <=, in [...] ,like %..., like ...% , like %...% , contain and not contain, with type detection based on field type. In the event of incompatibility, compile or code generation errors may be triggered. 
  if the filter condition is or/and relation,the expressions  should how to link . for example ,I add endpoint_http_cpm=from(Endpoint.*).filter(tags contain "http.method:GET" or tags contain "http.method:POST" ) . it is error . What should be the correct expression

**Comment by wu-sheng** [comment, marked as answer]
You could use filter as chain mode to support AND.
And contain accept an array, which means OR.

**Reply by BritneyStars** [reply]
OK! Thank you ! I have another confuse. I use ActiveSpan.tag("resultCode","1") to custom tags and add resultCode to searchableTracesTags in application.yml. When I define endpoint_cpm_result_code = from(Endpoint.).filter(tags contain "resultCode:1").cpm() in core.oal. But after runtime, I found the endpoint_cpm_result_code metric does not exist in metrics_cpm_xxx indice .why this? If I define service_cpm_http_other = from(Service.).filter(tags not contain "http.method:GET").cpm() , the metric is exist in metrics_cpm_xx indice . Customer tags can be used as a filter condition in .oal

**Reply by wu-sheng** [reply]
If `endpoint_cpm_result_code` doesn't exist, then there is nothing passed `filter`. 

For all custom tags, they should have been put into the `filter`.(L255)

https://github.com/apache/skywalking/blob/444adc6be1f6a1c0f6b628d7fcb8e23fe30d1529/oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/MultiScopesAnalysisListener.java#L243-L255

You could debug this line of codes and `org.apache.skywalking.oap.server.core.analysis.metrics.expression.ContainMatch` to see what is going on

**Reply by BritneyStars** [reply]
I debug the codes and found the span type is Local where the custom tags is located is ,So it does not enter the setPublicAttrs method in **skywalking/oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/listener/MultiScopesAnalysisListener.java** . it only enter parseLocal method and not record the Tags. 
![code](https://user-images.githubusercontent.com/26988781/131622400-1c18c6cc-eca7-4833-9bd4-2164d4bb5d88.png)
The method notifyLocalListener(spanObject, segmentObject) does not record the customer tags

**Reply by wu-sheng** [reply]
Oh, if you mean local span, no, it wouldn't be there. Because only entry span generates the source of service, which represents the traffic of service. And exit span generates tags for relationship(such as service calls service).

**Reply by wu-sheng** [reply]
Could you share what is your use case? If it is local span related, I would like to recommend logic span concept.

**Reply by BritneyStars** [reply]
I want to customize the business logic response status to some schedule task . For example, 
```
@Scheduled(cron="0 0/10 * * * ?")
public void fetchTaskSchedule(){
   .........
  Boolean flag=method();
  if(flag){
        ActiveSpan.tag("batch.resultCode","1");
	ActiveSpan.tag("batch.resultMsg","获取xxx成功");
  }else{
        ActiveSpan.tag("batch.resultCode","0");
	ActiveSpan.tag("batch.resultMsg","获取xxx失败");
  }
}
```
I want to calculate the count which batch.resultCode=0 or 1 by OAL . But because this is local span ,so it can not use the tag filter function .

**Reply by wu-sheng** [reply]
Then 2 things you could do.
1. You could contribute a pull request, make endpoint(source) generated by logic span including all other tags. 
2. Use manual API to tag this local span as logic span in your codes.
